python/ml4ir/base/model/calibrations/temperature_scaling.py

```diff
@@ -0,0 +1,247 @@
+from ml4ir.base.model.scoring.scoring_model import ScorerBase
+
+import shutil
+
+TEMPERATURE_SCALE = 'temp_scaling_scores'
+
+
+def make_val_and_grad_fn(fn):
+    """ Returns a function that computes y = fn(*args, **kwargs) and gradients of y wrt
+    each of args and  kwargs
+    parameters:
+        fn : Python callable to be differentiated
+    """
+    @functools.wraps(fn)
+    def val_and_grad(x):
+        return tfp.math.value_and_gradient(fn, x)
+    return val_and_grad
+
+
+@contextlib.contextmanager
+def timed_execution(logger):
+    """ a context manager to compute the execution time of a process usually within
+     a 'with' statement
+    Parameter:
+        logger: logging.Logger object
+    """
+    t0 = time.time()
+    yield
+    dt = time.time() - t0
+    logger.info(f'Evaluation took: {dt} seconds')
+
+
+def np_value(tensor):
+    """Get numpy value out of possibly nested tuple of tensors.
+    Parameters:
+        tensor: TF.Tensor
+                a tensor object
+    """
+    if isinstance(tensor, tuple):
+        return type(tensor)(*(np_value(t) for t in tensor))
+    else:
+        return tensor.numpy()
+
+
+def run(optimizer, logger):
+    """Run an optimizer and measure it's evaluation time."""
+    optimizer()  # Warmup.
+    with timed_execution(logger):
+        result = optimizer()
+    return np_value(result)
+
+
+def accuracy(scores, labels):
+    """
+    returns accuracy
+    Parameters:
+        scores: tf.Tensor object that contains softmaxes
+        labels: tf.Tensor object that contains labels
+
+    """
+    acc = tf.keras.metrics.Accuracy()
+    acc.update_state(labels, tf.argmax(scores, axis=-1))
+    return acc.result().numpy()
+
+
+def dict_to_zipped_csv(data_dic, root_dir):
+    """
+    saves input dictionary to a zipped csv file
+    Parameters:
+        data_dic: input dict to be converted to  zipped csv file
+        path: path to save the output file
+    """
+    # creating zip dir
+    zip_dir_path = os.path.join(root_dir, TEMPERATURE_SCALE)
+    os.mkdir(zip_dir_path)
+    csv_path = os.path.join(zip_dir_path, f'{TEMPERATURE_SCALE}.csv')
+
+    # creating .csv & .zip files
+    pd.DataFrame.from_dict(data_dic).to_csv(csv_path, index=False)
+    shutil.make_archive(zip_dir_path, "zip", root_dir, TEMPERATURE_SCALE)
+
+    # removing the dir, keeping only the zip
+    shutil.rmtree(zip_dir_path)
+    return zip_dir_path
+
+
+def get_intermediate_model(model, scorer):
+    """
+    returns a tf.keras.models.Model copy of the relavence_model.model. This model must generate
+    logits (inputs of softmax).
+    Parameters:
+        model:
+        scorer:
+    """
+    # get  last layer's output  --> MUST **NOT** BE AN ACTIVATION (e.g. SOFTMAX) LAYER
+    final_layer_name = scorer.model_config['layers'][-1]['name']
+    layer_output = model.get_layer(name=final_layer_name).output
+
+    return tf.keras.models.Model(inputs=model.input, outputs=layer_output)
```

Comment:
@marjanhs One thing I'm concerned about and should be covered in the unittests is you are not freezing any of the weights before you calibrate the model. Every weight will be relearned/adjusted in this case.

Reply:
I don't think that can cause a problem as we are creating `tf.constant` from logits  in `nll_loss_fn` https://github.com/salesforce/ml4ir/pull/115/files#diff-5e55239ed1f122b6b40beccad8e8301108acd2e382474342814c595590f776dcR177

Doesn't that freeze the weights ?

Reply:
Yes, I noticed that after I left this comments. Forgot to update this. I think your current implementation works because you use `tf.constant` on the logits. But I fear it is not going to be very optimal on a large dataset. I recommend leaving it as is for now and we should sync up about effort needed to use `RelevanceModel` directly with freezing all the learned weights and then modifying the last layer with temperature. This has a lot of advantages like directly saving the TFRecord signature with the temperature scaling embedded into it. We can come back to this though.